### PR TITLE
add repo image size and date

### PR DESCRIPTION
**Pull request by depombo**
Most of this PR revolves around changing the list private repository images client call to the same one used for public repository images which has a different data structure

### CODE REVIEW

File: src/modules/aws_ecr/index.ts [hunk 1]
```diff
@@ -31,16 +31,18 @@ class RepositoryImageMapper extends MapperBase<RepositoryImage> {
...
 
     // id is generated with imageDigest + tag
-    if (!image.imageId?.imageDigest) {
+    if (!image.imageDigest) {
       throw new Error('Invalid repository image');
     }
-    out.imageDigest = image.imageId.imageDigest;
-    out.imageTag = image.imageId.imageTag ?? '<untagged>';
-
+    const imageTag = image.imageTags && image.imageTags.length > 0 ? image.imageTags[0] : '<untagged>';
```
Comment (dfellis) [inline src/modules/aws_ecr/index.ts:41]: You may be able to replace this with:

```suggestion
    const imageTag = image?.imageTags?.[0] ?? '<untagged>';
```

File: src/modules/aws_ecr/index.ts [hunk 2]
```diff
@@ -31,16 +31,18 @@ class RepositoryImageMapper extends MapperBase<RepositoryImage> {
...
+    out.imageDigest = image.imageDigest;
+    out.imageTag = imageTag;
+    if (image.imageSizeInBytes) out.sizeInMB = Math.round(image.imageSizeInBytes * 0.000001);
```
Comment (dfellis) [inline src/modules/aws_ecr/index.ts:44]: Shouldn't this be:

```suggestion
    if (image.imageSizeInBytes) out.sizeInMB = Math.round(image.imageSizeInBytes / 1024 / 1024);
```

**Review by dfellis — APPROVED** [review]

File: src/modules/aws_ecr/entity/repository_image.ts [hunk 3]
```diff
@@ -62,6 +62,26 @@ export class RepositoryImage {
...
+
+  /**
+   * @public
+   * Size of the image in MB
+   */
+  @Column({
+    nullable: true,
+    type: 'int',
+  })
+  sizeInMB: number;
```
Comment (mtp1376) [inline src/modules/aws_ecr/entity/repository_image.ts:83]: Not important for this PR, but maybe we should define a convention for units? (e.g. saving them always in MB)

File: src/modules/aws_ecr/index.ts [hunk 4]
```diff
@@ -115,13 +113,13 @@ class RepositoryImageMapper extends MapperBase<RepositoryImage> {
...
-    if (images && images.imageIds && images.imageIds.length > 0) {
-      await this.deleteRepositoryImage(
-        client,
-        images.imageIds,
-        repository.repositoryName,
-        repository.registryId,
-      );
+    if (images && images.imageDetails) {
+      for (const image of images.imageDetails) {
+        if (image.imageDigest && image.imageTags && image.imageTags.length > 0) {
```
Comment (mtp1376) [inline src/modules/aws_ecr/index.ts:118]: So the `image.imageTags` should always be set here? Is that a must? I see in the `repositoryImageMapper` it's being treated as possibly-undefined. But here we're relying on its existence. Just a non-blocking question.

File: src/modules/aws_ecr/index.ts [hunk 5]
```diff
@@ -158,9 +156,10 @@ class RepositoryImageMapper extends MapperBase<RepositoryImage> {
           const idRegion = decoded[4];
           if (enabledRegions.includes(idRegion)) {
             const client = (await ctx.getAwsClient(idRegion)) as AWS;
-            const rawImage = await this.getRepositoryImage(client.ecrClient, [imageId], decoded[3]);
-            if (rawImage?.images && rawImage.images[0]) {
-              return await this.repositoryImageMapper(rawImage.images[0], ctx, type, idRegion);
+            const rawImage = await this.listRepositoryImages(client.ecrClient, [imageId], decoded[3]);
```
Comment (mtp1376) [inline src/modules/aws_ecr/index.ts:159]: Maybe this `decoded` variable can be replaced by using `this.idFields` function? Or can be done as a follow-up refactor task.

File: src/modules/aws_ecr/index.ts [hunk 6]
```diff
@@ -200,14 +190,24 @@ class RepositoryImageMapper extends MapperBase<RepositoryImage> {
...
+                              ctx,
+                              'private',
+                              region,
+                            ),
+                          );
+                        }
+                      } else {
+                        out.push(await this.repositoryImageMapper(imageDetail, ctx, 'private', region));
+                      }
+                    }
```
Comment (mtp1376) [inline src/modules/aws_ecr/index.ts:210]: It's a bit hard to follow this nested code. Maybe we can wrap it in a function and early-return instead of `if-else` conditions?

File: test/modules/aws-ecr-integration.ts [hunk 7]
```diff
@@ -203,7 +203,14 @@ describe('ECR Integration Testing', () => {
     `,
         (res: any[]) => {
           expect(res.length).toBe(3);
+          expect(res.map(i => i['size_in_mb']).reduce((a, b) => a + b, 0)).toBe(0);
           expect(res.filter(i => i['image_tag'] === '<untagged>').length).toBe(1);
+          const now = new Date();
+          const pushedAt = new Date(res[0]['pushed_at']);
+          expect(pushedAt.getDate()).toBe(now.getDate());
```
Comment (yrobla) [inline test/modules/aws-ecr-integration.ts:210]: isn't there a way to just do it on single line?

**Review by yrobla — COMMENTED** [review]
added comment

File: test/modules/aws-ecr-integration.ts [hunk 8]
```diff
@@ -203,7 +203,14 @@ describe('ECR Integration Testing', () => {
     `,
         (res: any[]) => {
           expect(res.length).toBe(3);
+          expect(res.map(i => i['size_in_mb']).reduce((a, b) => a + b, 0)).toBe(0);
```
Comment (aguillenv) [inline test/modules/aws-ecr-integration.ts:206]: either we are pushing something wrong to the repo or this check is not working properly, the size should not be 0, I think. If you see [here](https://gallery.ecr.aws/docker/library/hello-world) the image tags tab it says 122MB

File: src/modules/aws_ecr/index.ts [hunk 9]
```diff
@@ -2,7 +2,7 @@ import {
   ECR,
   Repository as RepositoryAws,
   paginateDescribeRepositories,
-  Image as ImageAws,
+  ImageDetail,
```
Comment (aguillenv) [inline src/modules/aws_ecr/index.ts:5]: is this a superset if `Image`?

File: src/modules/aws_ecr/index.ts [hunk 10]
```diff
@@ -31,16 +31,18 @@ class RepositoryImageMapper extends MapperBase<RepositoryImage> {
...
+    if (!image.imageDigest) {
       throw new Error('Invalid repository image');
     }
-    out.imageDigest = image.imageId.imageDigest;
-    out.imageTag = image.imageId.imageTag ?? '<untagged>';
-
+    const imageTag = image.imageTags?.[0] ?? '<untagged>';
+    out.imageDigest = image.imageDigest;
+    out.imageTag = imageTag;
+    if (image.imageSizeInBytes) out.sizeInMB = Math.round(image.imageSizeInBytes / 1024 / 1024);
```
Comment (aguillenv) [inline src/modules/aws_ecr/index.ts:44]: Maybe we should make sure `image.imageSizeInBytes` is a number? I think this would do the work
```suggestion
    if (image.imageSizeInBytes) out.sizeInMB = Math.round(+image.imageSizeInBytes / 1024 / 1024);
```

File: src/modules/aws_ecr/index.ts [hunk 11]
```diff
@@ -115,13 +113,13 @@ class RepositoryImageMapper extends MapperBase<RepositoryImage> {
...
     );
-    if (images && images.imageIds && images.imageIds.length > 0) {
-      await this.deleteRepositoryImage(
-        client,
-        images.imageIds,
-        repository.repositoryName,
-        repository.registryId,
-      );
+    if (images && images.imageDetails) {
+      for (const image of images.imageDetails) {
```
Comment (aguillenv) [inline src/modules/aws_ecr/index.ts:117]: nit:
```suggestion
    for (const image of images?.imageDetails ?? []) {
```

File: test/modules/aws-ecr-integration.ts (hunk 7 above)
Comment (depombo) [reply test/modules/aws-ecr-integration.ts:210]: It is on the same date, but at a different time. The only way I can think of is to get both dates as an ISO string and truncate them to exclude the time, but that would be harder to follow even if it involves less LOC. I could check that `now` is bigger than `pushed_at` in a single line, but that wouldn't be as exhaustive.

File: src/modules/aws_ecr/index.ts (hunk 9 above)
Comment (depombo) [reply src/modules/aws_ecr/index.ts:5]: not a straight super set, there are some minor differences between the two

File: test/modules/aws-ecr-integration.ts (hunk 8 above)
Comment (depombo) [reply test/modules/aws-ecr-integration.ts:206]: size is 14kB in the link you provided though I am seeing 4kB, but both of those round to 0 in MB. I think it is okay as-is or should do you want to track kB or bytes? That will be too small when the image is in GB

File: src/modules/aws_ecr/index.ts (hunk 10 above)
Comment (depombo) [reply src/modules/aws_ecr/index.ts:44]: it is typed into a number

File: src/modules/aws_ecr/index.ts (hunk 5 above)
Comment (depombo) [reply src/modules/aws_ecr/index.ts:159]: seems like more of a nit, can be a follow-up. I think it is okay as-is

File: src/modules/aws_ecr/entity/repository_image.ts (hunk 3 above)
Comment (depombo) [reply src/modules/aws_ecr/entity/repository_image.ts:83]: I believe it is MB everywhere

File: test/modules/aws-ecr-integration.ts (hunk 8 above)
Comment (aguillenv) [reply test/modules/aws-ecr-integration.ts:206]: ohh I see, then that's ok, let's keep it as is!

**Review by aguillenv — APPROVED** [review]

**Review by mtp1376 — APPROVED** [review]
